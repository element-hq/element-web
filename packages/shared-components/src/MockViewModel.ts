/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { type ViewModel } from "./ViewModel";

/**
 * A mock view model that returns a static snapshot passed in the constructor, with no updates.
 */
export class MockViewModel<T> implements ViewModel<T> {
    public constructor(private snapshot: T) {}

    public getSnapshot = (): T => {
        return this.snapshot;
    };

    public subscribe(listener: () => void): () => void {
        return () => undefined;
    }
}
