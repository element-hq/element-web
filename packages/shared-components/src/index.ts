/*
 * Copyright 2025 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

// Components
export * from "./audio/AudioPlayerView";
export * from "./audio/Clock";
export * from "./audio/PlayPauseButton";
export * from "./audio/SeekBar";
export * from "./avatar/AvatarWithDetails";
export * from "./event-tiles/TextualEventView";
export * from "./message-body/MediaBody";
export * from "./pill-input/Pill";
export * from "./pill-input/PillInput";
export * from "./rich-list/RichItem";
export * from "./rich-list/RichList";
export * from "./utils/Box";
export * from "./utils/Flex";

// Utils
export * from "./utils/i18n"
export * from "./utils/humanize"
export * from "./utils/DateUtils"
export * from "./utils/numbers"

// MVVM
export * from "./ViewWrapper"
export * from "./ViewModel"
export * from "./useViewModel"
export * from "./MockViewModel"

