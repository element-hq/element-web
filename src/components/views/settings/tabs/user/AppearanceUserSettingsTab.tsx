/*
Copyright 2019 New Vector Ltd
Copyright 2019 - 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { ChangeEvent, ReactNode } from "react";

import { _t } from "../../../../../languageHandler";
import SdkConfig from "../../../../../SdkConfig";
import SettingsStore from "../../../../../settings/SettingsStore";
import SettingsFlag from "../../../elements/SettingsFlag";
import Field from "../../../elements/Field";
import AccessibleButton from "../../../elements/AccessibleButton";
import { SettingLevel } from "../../../../../settings/SettingLevel";
import { UIFeature } from "../../../../../settings/UIFeature";
import { Layout } from "../../../../../settings/enums/Layout";
import LayoutSwitcher from "../../LayoutSwitcher";
import FontScalingPanel from "../../FontScalingPanel";
import ThemeChoicePanel from "../../ThemeChoicePanel";
import ImageSizePanel from "../../ImageSizePanel";
import SettingsTab from "../SettingsTab";
import { SettingsSection } from "../../shared/SettingsSection";
import SettingsSubsection from "../../shared/SettingsSubsection";
import MatrixClientContext from "../../../../../contexts/MatrixClientContext";

interface IProps {}

interface IState {
    useBundledEmojiFont: boolean;
    useSystemFont: boolean;
    systemFont: string;
    showAdvanced: boolean;
    layout: Layout;
    // User profile data for the message preview
    userId?: string;
    displayName?: string;
    avatarUrl?: string;
}

export default class AppearanceUserSettingsTab extends React.Component<IProps, IState> {
    public static contextType = MatrixClientContext;
    public context!: React.ContextType<typeof MatrixClientContext>;

    private readonly MESSAGE_PREVIEW_TEXT = _t("common|preview_message");

    private unmounted = false;

    public constructor(props: IProps) {
        super(props);

        this.state = {
            useBundledEmojiFont: SettingsStore.getValue("useBundledEmojiFont"),
            useSystemFont: SettingsStore.getValue("useSystemFont"),
            systemFont: SettingsStore.getValue("systemFont"),
            showAdvanced: false,
            layout: SettingsStore.getValue("layout"),
        };
    }

    public async componentDidMount(): Promise<void> {
        // Fetch the current user profile for the message preview
        const client = this.context;
        const userId = client.getUserId()!;
        const profileInfo = await client.getProfileInfo(userId);
        if (this.unmounted) return;

        this.setState({
            userId,
            displayName: profileInfo.displayname,
            avatarUrl: profileInfo.avatar_url,
        });
    }

    public componentWillUnmount(): void {
        this.unmounted = true;
    }

    private onLayoutChanged = (layout: Layout): void => {
        this.setState({ layout: layout });
    };

    private renderAdvancedSection(): ReactNode {
        if (!SettingsStore.getValue(UIFeature.AdvancedSettings)) return null;

        const brand = SdkConfig.get().brand;
        const toggle = (
            <AccessibleButton
                kind="link"
                onClick={() => this.setState({ showAdvanced: !this.state.showAdvanced })}
                aria-expanded={this.state.showAdvanced}
            >
                {this.state.showAdvanced ? _t("action|hide_advanced") : _t("action|show_advanced")}
            </AccessibleButton>
        );

        let advanced: React.ReactNode;

        if (this.state.showAdvanced) {
            const tooltipContent = _t("settings|appearance|custom_font_description", { brand });
            advanced = (
                <>
                    <SettingsFlag name="useCompactLayout" level={SettingLevel.DEVICE} useCheckbox={true} />

                    <SettingsFlag
                        name="useBundledEmojiFont"
                        level={SettingLevel.DEVICE}
                        useCheckbox={true}
                        onChange={(checked) => this.setState({ useBundledEmojiFont: checked })}
                    />
                    <SettingsFlag
                        name="useSystemFont"
                        level={SettingLevel.DEVICE}
                        useCheckbox={true}
                        onChange={(checked) => this.setState({ useSystemFont: checked })}
                    />
                    <Field
                        className="mx_AppearanceUserSettingsTab_checkboxControlledField"
                        label={SettingsStore.getDisplayName("systemFont")!}
                        onChange={(value: ChangeEvent<HTMLInputElement>) => {
                            this.setState({
                                systemFont: value.target.value,
                            });

                            SettingsStore.setValue("systemFont", null, SettingLevel.DEVICE, value.target.value);
                        }}
                        tooltipContent={tooltipContent}
                        forceTooltipVisible={true}
                        disabled={!this.state.useSystemFont}
                        value={this.state.systemFont}
                    />
                </>
            );
        }
        return (
            <SettingsSubsection>
                {toggle}
                {advanced}
            </SettingsSubsection>
        );
    }

    public render(): React.ReactNode {
        return (
            <SettingsTab data-testid="mx_AppearanceUserSettingsTab">
                <SettingsSection>
                    <ThemeChoicePanel />
                    <LayoutSwitcher
                        userId={this.state.userId}
                        displayName={this.state.displayName}
                        avatarUrl={this.state.avatarUrl}
                        messagePreviewText={this.MESSAGE_PREVIEW_TEXT}
                        onLayoutChanged={this.onLayoutChanged}
                    />
                    <FontScalingPanel />
                    {this.renderAdvancedSection()}
                    <ImageSizePanel />
                </SettingsSection>
            </SettingsTab>
        );
    }
}
