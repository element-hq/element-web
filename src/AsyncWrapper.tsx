/*
Copyright 2024 New Vector Ltd.
Copyright 2015-2021 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React, { ComponentType, PropsWithChildren } from "react";
import { logger } from "matrix-js-sdk/src/logger";

import { _t } from "./languageHandler";
import BaseDialog from "./components/views/dialogs/BaseDialog";
import DialogButtons from "./components/views/elements/DialogButtons";
import Spinner from "./components/views/elements/Spinner";

type AsyncImport<T> = { default: T };

interface IProps {
    // A promise which resolves with the real component
    prom: Promise<ComponentType<any> | AsyncImport<ComponentType<any>>>;
    onFinished(): void;
}

interface IState {
    component?: ComponentType<PropsWithChildren<any>>;
    error?: Error;
}

/**
 * Wrap an asynchronous loader function with a react component which shows a
 * spinner until the real component loads.
 */
export default class AsyncWrapper extends React.Component<IProps, IState> {
    private unmounted = false;

    public state: IState = {};

    public componentDidMount(): void {
        this.props.prom
            .then((result) => {
                if (this.unmounted) return;

                // Take the 'default' member if it's there, then we support
                // passing in just an import()ed module, since ES6 async import
                // always returns a module *namespace*.
                const component = (result as AsyncImport<ComponentType>).default
                    ? (result as AsyncImport<ComponentType>).default
                    : (result as ComponentType);
                this.setState({ component });
            })
            .catch((e) => {
                logger.warn("AsyncWrapper promise failed", e);
                this.setState({ error: e });
            });
    }

    public componentWillUnmount(): void {
        this.unmounted = true;
    }

    private onWrapperCancelClick = (): void => {
        this.props.onFinished();
    };

    public render(): React.ReactNode {
        if (this.state.component) {
            const Component = this.state.component;
            return <Component {...this.props} />;
        } else if (this.state.error) {
            return (
                <BaseDialog onFinished={this.props.onFinished} title={_t("common|error")}>
                    {_t("failed_load_async_component")}
                    <DialogButtons
                        primaryButton={_t("action|dismiss")}
                        onPrimaryButtonClick={this.onWrapperCancelClick}
                        hasCancel={false}
                    />
                </BaseDialog>
            );
        } else {
            // show a spinner until the component is loaded.
            return <Spinner />;
        }
    }
}
