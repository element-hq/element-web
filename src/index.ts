/*
Copyright 2024 New Vector Ltd.
Copyright 2019 The Matrix.org Foundation C.I.C.
Copyright 2015, 2016 OpenMarket Ltd

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

// Import the js-sdk so the proper `request` object can be set. This does some
// magic with the browser injection to make all subsequent imports work fine.
import "matrix-js-sdk/src/browser-index";
