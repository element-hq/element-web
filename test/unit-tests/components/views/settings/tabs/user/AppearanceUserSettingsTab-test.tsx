/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import { render } from "jest-matrix-react";
import React from "react";
import { MatrixClient } from "matrix-js-sdk/src/matrix";

import AppearanceUserSettingsTab from "../../../../../../../src/components/views/settings/tabs/user/AppearanceUserSettingsTab";
import { withClientContextRenderOptions, stubClient } from "../../../../../../test-utils";

describe("AppearanceUserSettingsTab", () => {
    let client: MatrixClient;

    beforeEach(() => {
        client = stubClient();
    });

    it("should render", () => {
        const { asFragment } = render(<AppearanceUserSettingsTab />, withClientContextRenderOptions(client));
        expect(asFragment()).toMatchSnapshot();
    });
});
