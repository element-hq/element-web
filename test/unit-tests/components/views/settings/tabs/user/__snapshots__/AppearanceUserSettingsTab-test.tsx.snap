// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppearanceUserSettingsTab should render 1`] = `
<DocumentFragment>
  <div
    class="mx_SettingsTab"
    data-testid="mx_AppearanceUserSettingsTab"
  >
    <div
      class="mx_SettingsTab_sections"
    >
      <div
        class="mx_SettingsSection"
      >
        <div
          class="mx_SettingsSection_subSections"
        >
          <div
            class="mx_SettingsSubsection mx_SettingsSubsection_newUi"
            data-testid="themePanel"
          >
            <div
              class="mx_SettingsSubsectionHeading"
            >
              <h3
                class="mx_Heading_h3 mx_SettingsSubsectionHeading_heading"
              >
                Theme
              </h3>
            </div>
            <div
              class="mx_SettingsSubsection_content mx_SettingsSubsection_content_newUi"
            >
              <form
                class="_root_dgy0u_24 mx_ThemeChoicePanel_ThemeSelectors"
              >
                <div
                  class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector mx_ThemeChoicePanel_themeSelector_disabled cpd-theme-light"
                >
                  <div
                    class="_inline-field-control_dgy0u_52"
                  >
                    <div
                      class="_container_1vw5h_18"
                    >
                      <input
                        class="_input_1vw5h_26"
                        disabled=""
                        id="radix-:r0:"
                        name="themeSelector"
                        title=""
                        type="radio"
                        value="light"
                      />
                      <div
                        class="_ui_1vw5h_27"
                      />
                    </div>
                  </div>
                  <div
                    class="_inline-field-body_dgy0u_46"
                  >
                    <label
                      class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
                      for="radix-:r0:"
                    >
                      Light
                    </label>
                  </div>
                </div>
                <div
                  class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector mx_ThemeChoicePanel_themeSelector_disabled cpd-theme-dark"
                >
                  <div
                    class="_inline-field-control_dgy0u_52"
                  >
                    <div
                      class="_container_1vw5h_18"
                    >
                      <input
                        class="_input_1vw5h_26"
                        disabled=""
                        id="radix-:r1:"
                        name="themeSelector"
                        title=""
                        type="radio"
                        value="dark"
                      />
                      <div
                        class="_ui_1vw5h_27"
                      />
                    </div>
                  </div>
                  <div
                    class="_inline-field-body_dgy0u_46"
                  >
                    <label
                      class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
                      for="radix-:r1:"
                    >
                      Dark
                    </label>
                  </div>
                </div>
                <div
                  class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector mx_ThemeChoicePanel_themeSelector_disabled cpd-theme-light"
                >
                  <div
                    class="_inline-field-control_dgy0u_52"
                  >
                    <div
                      class="_container_1vw5h_18"
                    >
                      <input
                        class="_input_1vw5h_26"
                        disabled=""
                        id="radix-:r2:"
                        name="themeSelector"
                        title=""
                        type="radio"
                        value="light-high-contrast"
                      />
                      <div
                        class="_ui_1vw5h_27"
                      />
                    </div>
                  </div>
                  <div
                    class="_inline-field-body_dgy0u_46"
                  >
                    <label
                      class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
                      for="radix-:r2:"
                    >
                      High contrast
                    </label>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="_separator_144s5_17"
              data-kind="primary"
              data-orientation="horizontal"
              role="separator"
            />
          </div>
          <div
            class="mx_SettingsSubsection mx_SettingsSubsection_newUi"
            data-testid="layoutPanel"
          >
            <div
              class="mx_SettingsSubsectionHeading"
            >
              <h3
                class="mx_Heading_h3 mx_SettingsSubsectionHeading_heading"
              >
                Message layout
              </h3>
            </div>
            <div
              class="mx_SettingsSubsection_content mx_SettingsSubsection_content_newUi"
            >
              <form
                class="_root_dgy0u_24 mx_LayoutSwitcher_LayoutSelector"
              >
                <div
                  class="_field_dgy0u_34 mxLayoutSwitcher_LayoutSelector_LayoutRadio"
                >
                  <label
                    aria-label="Modern"
                    class="_label_dgy0u_67"
                    for="radix-:r3:"
                  >
                    <div
                      class="mxLayoutSwitcher_LayoutSelector_LayoutRadio_inline"
                    >
                      <div
                        class="_container_1vw5h_18"
                      >
                        <input
                          checked=""
                          class="_input_1vw5h_26"
                          id="radix-:r3:"
                          name="layout"
                          title=""
                          type="radio"
                          value="group"
                        />
                        <div
                          class="_ui_1vw5h_27"
                        />
                      </div>
                      <span>
                        Modern
                      </span>
                    </div>
                    <hr
                      class="mxLayoutSwitcher_LayoutSelector_LayoutRadio_separator"
                    />
                    <div
                      class="mxLayoutSwitcher_LayoutSelector_LayoutRadio_EventTilePreview"
                      role="presentation"
                    >
                      <div
                        aria-atomic="true"
                        aria-live="off"
                        class="mx_EventTile"
                        data-event-id="$9999999999999999999999999999999999999999999"
                        data-has-reply="false"
                        data-layout="group"
                        data-scroll-tokens="$9999999999999999999999999999999999999999999"
                        data-self="true"
                        tabindex="-1"
                      >
                        <div
                          class="mx_DisambiguatedProfile"
                        >
                          <span
                            class="mx_Username_color2 mx_DisambiguatedProfile_displayName"
                            dir="auto"
                          >
                            @userId:matrix.org
                          </span>
                        </div>
                        <div
                          class="mx_EventTile_avatar"
                        >
                          <span
                            class="_avatar_mcap2_17 mx_BaseAvatar _avatar-imageless_mcap2_61"
                            data-color="2"
                            data-testid="avatar-img"
                            data-type="round"
                            role="presentation"
                            style="--cpd-avatar-size: 30px;"
                            title="@userId:matrix.org"
                          >
                            u
                          </span>
                        </div>
                        <div
                          class="mx_EventTile_line"
                        >
                          <div
                            class="mx_MTextBody mx_EventTile_content"
                          >
                            <div
                              class="mx_EventTile_body translate"
                              dir="auto"
                            >
                              Hey you. You're the best!
                            </div>
                          </div>
                          <div
                            aria-label="Message Actions"
                            aria-live="off"
                            class="mx_MessageActionBar"
                            role="toolbar"
                          >
                            <div
                              aria-label="Edit"
                              class="mx_AccessibleButton mx_MessageActionBar_iconButton"
                              role="button"
                              tabindex="0"
                            >
                              <div />
                            </div>
                            <div
                              aria-expanded="false"
                              aria-haspopup="true"
                              aria-label="Options"
                              class="mx_AccessibleButton mx_MessageActionBar_iconButton mx_MessageActionBar_optionsButton"
                              role="button"
                              tabindex="-1"
                            >
                              <svg
                                fill="currentColor"
                                height="1em"
                                viewBox="0 0 24 24"
                                width="1em"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M6 14c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 4 12c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 6 10c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412A1.926 1.926 0 0 1 6 14Zm6 0c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 10 12c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 12 10c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412A1.926 1.926 0 0 1 12 14Zm6 0c-.55 0-1.02-.196-1.413-.588A1.926 1.926 0 0 1 16 12c0-.55.196-1.02.587-1.412A1.926 1.926 0 0 1 18 10c.55 0 1.02.196 1.413.588.391.391.587.862.587 1.412 0 .55-.196 1.02-.587 1.412A1.926 1.926 0 0 1 18 14Z"
                                />
                              </svg>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </label>
                </div>
                <div
                  class="_field_dgy0u_34 mxLayoutSwitcher_LayoutSelector_LayoutRadio"
                >
                  <label
                    aria-label="Message bubbles"
                    class="_label_dgy0u_67"
                    for="radix-:rc:"
                  >
                    <div
                      class="mxLayoutSwitcher_LayoutSelector_LayoutRadio_inline"
                    >
                      <div
                        class="_container_1vw5h_18"
                      >
                        <input
                          class="_input_1vw5h_26"
                          id="radix-:rc:"
                          name="layout"
                          title=""
                          type="radio"
                          value="bubble"
                        />
                        <div
                          class="_ui_1vw5h_27"
                        />
                      </div>
                      <span>
                        Message bubbles
                      </span>
                    </div>
                    <hr
                      class="mxLayoutSwitcher_LayoutSelector_LayoutRadio_separator"
                    />
                    <div
                      class="mxLayoutSwitcher_LayoutSelector_LayoutRadio_EventTilePreview"
                      role="presentation"
                    >
                      <div
                        aria-atomic="true"
                        aria-live="off"
                        class="mx_EventTile"
                        data-event-id="$9999999999999999999999999999999999999999999"
                        data-has-reply="false"
                        data-layout="bubble"
                        data-scroll-tokens="$9999999999999999999999999999999999999999999"
                        data-self="true"
                        tabindex="-1"
                      >
                        <div
                          class="mx_DisambiguatedProfile"
                        >
                          <span
                            class="mx_Username_color2 mx_DisambiguatedProfile_displayName"
                            dir="auto"
                          >
                            @userId:matrix.org
                          </span>
                        </div>
                        <div
                          class="mx_EventTile_avatar"
                        >
                          <span
                            class="_avatar_mcap2_17 mx_BaseAvatar _avatar-imageless_mcap2_61"
                            data-color="2"
                            data-testid="avatar-img"
                            data-type="round"
                            role="presentation"
                            style="--cpd-avatar-size: 30px;"
                            title="@userId:matrix.org"
                          >
                            u
                          </span>
                        </div>
                        <div
                          class="mx_EventTile_line"
                        >
                          <div
                            class="mx_MTextBody mx_EventTile_content"
                          >
                            <div
                              class="mx_EventTile_body translate"
                              dir="auto"
                            >
                              Hey you. You're the best!
                            </div>
                          </div>
                          <div
                            aria-label="Message Actions"
                            aria-live="off"
                            class="mx_MessageActionBar"
                            role="toolbar"
                          >
                            <div
                              aria-label="Edit"
                              class="mx_AccessibleButton mx_MessageActionBar_iconButton"
                              role="button"
                              tabindex="0"
                            >
                              <div />
                            </div>
                            <div
                              aria-expanded="false"
                              aria-haspopup="true"
                              aria-label="Options"
                              class="mx_AccessibleButton mx_MessageActionBar_iconButton mx_MessageActionBar_optionsButton"
                              role="button"
                              tabindex="-1"
                            >
                              <svg
                                fill="currentColor"
                                height="1em"
                                viewBox="0 0 24 24"
                                width="1em"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M6 14c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 4 12c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 6 10c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412A1.926 1.926 0 0 1 6 14Zm6 0c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 10 12c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 12 10c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412A1.926 1.926 0 0 1 12 14Zm6 0c-.55 0-1.02-.196-1.413-.588A1.926 1.926 0 0 1 16 12c0-.55.196-1.02.587-1.412A1.926 1.926 0 0 1 18 10c.55 0 1.02.196 1.413.588.391.391.587.862.587 1.412 0 .55-.196 1.02-.587 1.412A1.926 1.926 0 0 1 18 14Z"
                                />
                              </svg>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </label>
                </div>
                <div
                  class="_field_dgy0u_34 mxLayoutSwitcher_LayoutSelector_LayoutRadio"
                >
                  <label
                    aria-label="IRC (experimental)"
                    class="_label_dgy0u_67"
                    for="radix-:rl:"
                  >
                    <div
                      class="mxLayoutSwitcher_LayoutSelector_LayoutRadio_inline"
                    >
                      <div
                        class="_container_1vw5h_18"
                      >
                        <input
                          class="_input_1vw5h_26"
                          id="radix-:rl:"
                          name="layout"
                          title=""
                          type="radio"
                          value="irc"
                        />
                        <div
                          class="_ui_1vw5h_27"
                        />
                      </div>
                      <span>
                        IRC (experimental)
                      </span>
                    </div>
                    <hr
                      class="mxLayoutSwitcher_LayoutSelector_LayoutRadio_separator"
                    />
                    <div
                      class="mxLayoutSwitcher_LayoutSelector_LayoutRadio_EventTilePreview mx_IRCLayout"
                      role="presentation"
                    >
                      <div
                        aria-atomic="true"
                        aria-live="off"
                        class="mx_EventTile"
                        data-event-id="$9999999999999999999999999999999999999999999"
                        data-has-reply="false"
                        data-layout="irc"
                        data-scroll-tokens="$9999999999999999999999999999999999999999999"
                        data-self="true"
                        tabindex="-1"
                      >
                        <div
                          class="mx_DisambiguatedProfile"
                        >
                          <span
                            class="mx_Username_color2 mx_DisambiguatedProfile_displayName"
                            dir="auto"
                          >
                            @userId:matrix.org
                          </span>
                        </div>
                        <div
                          class="mx_EventTile_avatar"
                        >
                          <span
                            class="_avatar_mcap2_17 mx_BaseAvatar _avatar-imageless_mcap2_61"
                            data-color="2"
                            data-testid="avatar-img"
                            data-type="round"
                            role="presentation"
                            style="--cpd-avatar-size: 14px;"
                            title="@userId:matrix.org"
                          >
                            u
                          </span>
                        </div>
                        <div
                          class="mx_EventTile_line"
                        >
                          <div
                            class="mx_MTextBody mx_EventTile_content"
                          >
                            <div
                              class="mx_EventTile_body translate"
                              dir="auto"
                            >
                              Hey you. You're the best!
                            </div>
                          </div>
                          <div
                            aria-label="Message Actions"
                            aria-live="off"
                            class="mx_MessageActionBar"
                            role="toolbar"
                          >
                            <div
                              aria-label="Edit"
                              class="mx_AccessibleButton mx_MessageActionBar_iconButton"
                              role="button"
                              tabindex="0"
                            >
                              <div />
                            </div>
                            <div
                              aria-expanded="false"
                              aria-haspopup="true"
                              aria-label="Options"
                              class="mx_AccessibleButton mx_MessageActionBar_iconButton mx_MessageActionBar_optionsButton"
                              role="button"
                              tabindex="-1"
                            >
                              <svg
                                fill="currentColor"
                                height="1em"
                                viewBox="0 0 24 24"
                                width="1em"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M6 14c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 4 12c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 6 10c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412A1.926 1.926 0 0 1 6 14Zm6 0c-.55 0-1.02-.196-1.412-.588A1.926 1.926 0 0 1 10 12c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 12 10c.55 0 1.02.196 1.412.588.392.391.588.862.588 1.412 0 .55-.196 1.02-.588 1.412A1.926 1.926 0 0 1 12 14Zm6 0c-.55 0-1.02-.196-1.413-.588A1.926 1.926 0 0 1 16 12c0-.55.196-1.02.587-1.412A1.926 1.926 0 0 1 18 10c.55 0 1.02.196 1.413.588.391.391.587.862.587 1.412 0 .55-.196 1.02-.587 1.412A1.926 1.926 0 0 1 18 14Z"
                                />
                              </svg>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </label>
                </div>
              </form>
              <form
                class="_root_dgy0u_24"
              >
                <div
                  class="_inline-field_dgy0u_40"
                >
                  <div
                    class="_inline-field-control_dgy0u_52"
                  >
                    <div
                      class="_container_qnvru_18"
                    >
                      <input
                        aria-describedby="radix-:rv:"
                        class="_input_qnvru_32"
                        id="radix-:ru:"
                        name="compactLayout"
                        title=""
                        type="checkbox"
                      />
                      <div
                        class="_ui_qnvru_42"
                      />
                    </div>
                  </div>
                  <div
                    class="_inline-field-body_dgy0u_46"
                  >
                    <label
                      class="_label_dgy0u_67"
                      for="radix-:ru:"
                    >
                      Show compact text and messages
                    </label>
                    <span
                      class="_message_dgy0u_98 _help-message_dgy0u_104"
                      id="radix-:rv:"
                    >
                      Modern layout must be selected to use this feature.
                    </span>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="_separator_144s5_17"
              data-kind="primary"
              data-orientation="horizontal"
              role="separator"
            />
          </div>
          <div
            class="mx_SettingsSubsection"
            data-testid="mx_FontScalingPanel"
          >
            <div
              class="mx_SettingsSubsectionHeading"
            >
              <h3
                class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
              >
                Font size
              </h3>
            </div>
            <div
              class="mx_SettingsSubsection_content mx_SettingsSubsection_contentStretch"
            >
              <div
                class="mx_Field mx_Field_select mx_FontScalingPanel_Dropdown"
              >
                <select
                  id="mx_Field_1"
                  label="Font size"
                  placeholder="Font size"
                  type="text"
                >
                  <option
                    value="-7"
                  >
                    9
                  </option>
                  <option
                    value="-6"
                  >
                    10
                  </option>
                  <option
                    value="-5"
                  >
                    11
                  </option>
                  <option
                    value="-4"
                  >
                    12
                  </option>
                  <option
                    value="-3"
                  >
                    13
                  </option>
                  <option
                    value="-2"
                  >
                    14
                  </option>
                  <option
                    value="-1"
                  >
                    15
                  </option>
                  <option
                    value="0"
                  >
                    16 (default)
                  </option>
                  <option
                    value="1"
                  >
                    17
                  </option>
                  <option
                    value="2"
                  >
                    18
                  </option>
                  <option
                    value="4"
                  >
                    20
                  </option>
                  <option
                    value="6"
                  >
                    22
                  </option>
                  <option
                    value="8"
                  >
                    24
                  </option>
                  <option
                    value="10"
                  >
                    26
                  </option>
                  <option
                    value="12"
                  >
                    28
                  </option>
                  <option
                    value="14"
                  >
                    30
                  </option>
                  <option
                    value="16"
                  >
                    32
                  </option>
                  <option
                    value="18"
                  >
                    34
                  </option>
                  <option
                    value="20"
                  >
                    36
                  </option>
                </select>
                <label
                  for="mx_Field_1"
                >
                  Font size
                </label>
              </div>
              <div
                class="mx_FontScalingPanel_preview mx_EventTilePreview_loader"
              >
                <div
                  class="mx_Spinner"
                >
                  <div
                    aria-label="Loading…"
                    class="mx_Spinner_icon"
                    data-testid="spinner"
                    role="progressbar"
                    style="width: 32px; height: 32px;"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="mx_SettingsSubsection"
          >
            <div
              class="mx_SettingsSubsection_content mx_SettingsSubsection_noHeading"
            >
              <div
                aria-expanded="false"
                class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_link"
                role="button"
                tabindex="0"
              >
                Show advanced
              </div>
            </div>
          </div>
          <div
            class="mx_SettingsSubsection"
          >
            <div
              class="mx_SettingsSubsectionHeading"
            >
              <h3
                class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
              >
                Image size in the timeline
              </h3>
            </div>
            <div
              class="mx_SettingsSubsection_content"
            >
              <div
                class="mx_ImageSizePanel_radios"
              >
                <label>
                  <div
                    class="mx_ImageSizePanel_size mx_ImageSizePanel_sizeDefault"
                  />
                  <label
                    class="mx_StyledRadioButton mx_StyledRadioButton_enabled mx_StyledRadioButton_checked"
                  >
                    <input
                      checked=""
                      name="image_size"
                      type="radio"
                      value="normal"
                    />
                    <div>
                      <div />
                    </div>
                    <div
                      class="mx_StyledRadioButton_content"
                    >
                      Default
                    </div>
                    <div
                      class="mx_StyledRadioButton_spacer"
                    />
                  </label>
                </label>
                <label>
                  <div
                    class="mx_ImageSizePanel_size mx_ImageSizePanel_sizeLarge"
                  />
                  <label
                    class="mx_StyledRadioButton mx_StyledRadioButton_enabled"
                  >
                    <input
                      name="image_size"
                      type="radio"
                      value="large"
                    />
                    <div>
                      <div />
                    </div>
                    <div
                      class="mx_StyledRadioButton_content"
                    >
                      Large
                    </div>
                    <div
                      class="mx_StyledRadioButton_spacer"
                    />
                  </label>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
